# 合併地震資料及爬文資料

In [1]:
import numpy as np
import pandas as pd

In [2]:
df_earthquake = pd.read_csv("earthquake_data_0620.csv")
df_ptt = pd.read_csv("ptt.csv")

In [3]:
df_earthquake.columns = ['編號', 'Date_format', 'Longitude', 'Latitude', '規模', '深度', '位置', '震度']

In [4]:
df_earthquake["IsOfficialReport"] = 1

In [5]:
df_earthquake.head()

編號          Date_format  Longitude  Latitude   規模    深度  \
0  小區域有感地震  2020-06-17 08:06:39     120.56     23.50  3.1   8.3   
1       33  2020-06-17 07:17:00     121.98     24.62  4.7  60.8   
2  小區域有感地震  2020-06-15 10:42:24     121.63     24.33  3.1  19.1   
3  小區域有感地震  2020-06-15 02:15:40     121.66     24.39  3.6  35.0   
4  小區域有感地震  2020-06-15 02:10:08     121.62     24.40  4.3  40.7   

                                       位置   震度  IsOfficialReport  
0    嘉義縣政府東方  28.0  公里 (位於嘉義縣竹崎鄉)(嘉義縣竹崎鄉)  3.0                 1  
1  宜蘭縣政府東偏南方  24.9  公里 (位於臺灣東部海域)(臺灣東部海域)  3.0                 1  
2    花蓮縣政府北方  37.6  公里 (位於花蓮縣秀林鄉)(花蓮縣秀林鄉)  1.0                 1  
3  宜蘭縣政府南偏西方  39.6  公里 (位於宜蘭縣南澳鄉)(宜蘭縣南澳鄉)  1.0                 1  
4  宜蘭縣政府南偏西方  39.3  公里 (位於宜蘭縣南澳鄉)(宜蘭縣南澳鄉)  2.0                 1

In [6]:
df_ptt["IsOfficialReport"] = 0

In [7]:
df_ptt.head()

Unnamed: 0                     Author      Board         City  \
0           0          werqq (sandy0980)  Gossiping       Taipei   
1           1                  inmee (J)  Gossiping       Taipei   
2           2  MiLuDaiBoom (ⓂⓛⓁⓤⒹⓐⓛⒷⓞⓞⓂ)  Gossiping       Taipei   
3           3               somehog (低進)  Gossiping       Taipei   
4           4                WeiU (微，Ｕ文)  Gossiping  Tainan City   

                                             Content  \
0  嗨 大家早安\n\n小妹剛剛被地震震了一下\n\n不曉得被震這一下後\n\n莎莎會不會受傷呢...   
1  剛剛上下震了好大一下\n\n讓小妹嚇得花枝亂戰\n\n一時之間不曉得該怎麼辦\n\n究竟.....   
2  \n\n\n欸欸欸\n\n剛剛在搖欸\n\n阿\n\n小妹我還在睡覺\n\n當然沒穿衣服啊\...   
3                    為什麼我又沒收到國家警報了?\n手機明明滿格!\n幹~\n\n   
4                       \n\n妹子我喜歡妳！！\n\n但地震了快逃啊！\n\n   

                       Date               Full IP              IP  \
0  Thu Jun 11 09:44:31 2020   223.136.205.58 (臺灣)  223.136.205.58   
1  Thu Jun 11 09:44:05 2020   223.137.176.72 (臺灣)  223.137.176.72   
2  Thu Jun 11 09:43:23 2020   110.50.129.111 (臺灣)  110.50.129.111   
3  Thu Jun 11 09:42:46 2020    119.14.24.108 (臺灣)   119.14.24.108   
4  Thu Jun 11 09:41:42 2020   119.14.206.234 (臺灣)  119.14.206.234   

  Longitude and Latitude                    Mobile  \
0   經度：121.532　緯度：25.048          Chunghwa Telecom   
1   經度：121.532　緯度：25.048   Taiwan Academic Network   
2   經度：121.532　緯度：25.048   Taiwan Academic Network   
3   經度：121.532　緯度：25.048  Hoshin Multimedia Center   
4   經度：120.213　緯度：22.991  Hoshin Multimedia Center   

                                Nation  \
0  AS - Asia ( 亞洲 ) TW - Taiwan ( 台灣 )   
1  AS - Asia ( 亞洲 ) TW - Taiwan ( 台灣 )   
2  AS - Asia ( 亞洲 ) TW - Taiwan ( 台灣 )   
3  AS - Asia ( 亞洲 ) TW - Taiwan ( 台灣 )   
4  AS - Asia ( 亞洲 ) TW - Taiwan ( 台灣 )   

                                                Push                  Title  \
0  ['', '', '本網站已依台灣網站內容分級規定處理。此區域為限制級，未滿十八歲者不得瀏覽...       [問卦] 地震後愛莉莎莎沒事吧？   
1  ['', '', '本網站已依台灣網站內容分級規定處理。此區域為限制級，未滿十八歲者不得瀏覽...  [問卦] 地震第一時間要做什麼才是正常人？   
2  ['', '', '本網站已依台灣網站內容分級規定處理。此區域為限制級，未滿十八歲者不得瀏覽...       [問卦] 地震時沒穿衣服怎麼辦？   
3  ['', '', '本網站已依台灣網站內容分級規定處理。此區域為限制級，未滿十八歲者不得瀏覽...    [爆卦] 幹!地震 為什麼我又邊緣人了   
4  ['', '', '本網站已依台灣網站內容分級規定處理。此區域為限制級，未滿十八歲者不得瀏覽...               [爆卦] 地震！   

   YouWin  Longitude  Latitude          Date_format  IsOfficialReport  
0       0    121.532    25.048  2020-06-11 09:44:31                 0  
1       0    121.532    25.048  2020-06-11 09:44:05                 0  
2       0    121.532    25.048  2020-06-11 09:43:23                 0  
3       0    121.532    25.048  2020-06-11 09:42:46                 0  
4       0    120.213    22.991  2020-06-11 09:41:42                 0

## 合併檔案

In [8]:
df = pd.merge(df_earthquake, df_ptt,how='outer', on = ['Date_format', 'Longitude', 'Latitude', 'IsOfficialReport'], sort="Date_format")

df["TimeDifference"] = "NaN"

In [9]:
df.head()

編號          Date_format  Longitude  Latitude  規模  深度   位置  震度  \
0  NaN  2008-05-16 11:31:40    121.000    23.500 NaN NaN  NaN NaN   
1  NaN  2009-07-14 19:18:33    121.000    23.500 NaN NaN  NaN NaN   
2  NaN  2009-08-13 22:04:59    121.000    23.500 NaN NaN  NaN NaN   
3  NaN  2010-03-18 22:54:47    120.684    24.147 NaN NaN  NaN NaN   
4  NaN  2011-03-12 02:22:52    139.690    35.690 NaN NaN  NaN NaN   

   IsOfficialReport  Unnamed: 0  ...    \
0                 0       998.0  ...     
1                 0       997.0  ...     
2                 0       996.0  ...     
3                 0       995.0  ...     
4                 0       994.0  ...     

                                             Content  \
0  ※ 引述《makihara (maki)》之銘言：\n: 看新聞很多截肢的人...而且很多是...   
1  轉自SHARE板\n\n 作者  mstar (Wayne Su)             ...   
2                                         這是國內一位河圖洛書   
3  ※ 引述《GundamXX》之銘言：\n\n\n劣退或水桶都好，五樓不要那麼沒品，要是有你要...   
4  \n下午那時剛要搭新宿中央縣回家,那時的我在地下大概1~2層的底下\n\n突然間輕微的搖晃,...   

                       Date                              Full IP  \
0  Fri May 16 11:31:40 2008                                  NaN   
1  Tue Jul 14 19:18:33 2009   61.229.29.48         (05/13 10:53)   
2  Thu Aug 13 22:04:59 2009   61.20.131.104        (08/13 22:53)   
3  Thu Mar 18 22:54:47 2010   112.104.159.100      (03/18 22:58)   
4  Sat Mar 12 02:22:52 2011   119.107.220.15       (03/12 02:25)   

                IP Longitude and Latitude          Mobile  \
0     61.229.29.48   經度：121.000　緯度：23.500           HiNet   
1     61.229.29.48   經度：121.000　緯度：23.500           HiNet   
2    61.20.131.104   經度：121.000　緯度：23.500      FarEasTone   
3  112.104.159.100   經度：120.684　緯度：24.147  Digital United   
4   119.107.220.15   經度：139.690　緯度：35.690      au one net   

                                Nation     Push  \
0  AS - Asia ( 亞洲 ) TW - Taiwan ( 台灣 )  No Push   
1  AS - Asia ( 亞洲 ) TW - Taiwan ( 台灣 )  No Push   
2  AS - Asia ( 亞洲 ) TW - Taiwan ( 台灣 )  No Push   
3  AS - Asia ( 亞洲 ) TW - Taiwan ( 台灣 )  No Push   
4   AS - Asia ( 亞洲 ) JP - Japan ( 日本 )  No Push   

                            Title YouWin  
0      Re: [新聞] 四川地震「別鋸我的腿 我寧願自殺」    0.0  
1   Re: [問卦] 有沒有地震震度都用幾顆原子彈來表示的 …    0.0  
2  [爆卦] 為什麼87水災921地震88水災全部在尾數八的年?    0.0  
3         Re: [問卦] 有沒有明天早上會大地震的八卦    0.0  
4                  [爆卦] 大地震時在新宿的我    0.0  

[5 rows x 23 columns]

# Cleaning

In [10]:
df.iloc[100]["IsOfficialReport"]==1

True

In [11]:
earthquake_index = []
for i in range(len(df)):
    if df.iloc[i]["IsOfficialReport"]==1:
        if df.iloc[i+1]["YouWin"]==1: #在一起地震後確實有「你贏了」推文的那些文章 i.e. 地震後有地震文
            #df_new = df_new.append(df.iloc[i], ignore_index=True)
            earthquake_index.append(i)
len(earthquake_index)

91

代表有91起地震是在其後有「你贏了」推文的

In [12]:
df.iloc[earthquake_index].head()

編號          Date_format  Longitude  Latitude   規模     深度  \
4959  小區域有感地震  2019-05-08 23:48:47     122.49     24.82  5.1  122.7   
4965  小區域有感地震  2019-05-10 21:49:30     121.92     24.50  3.9   40.4   
4968       36  2019-05-13 16:55:21     121.90     24.75  5.2   88.0   
5013  小區域有感地震  2019-05-19 00:38:34     122.19     24.03  4.7   22.5   
5025       37  2019-05-22 09:37:24     121.64     23.96  4.8   28.8   

                                        位置   震度  IsOfficialReport  Unnamed: 0  \
4959  宜蘭縣政府東方  74.2  公里 (位於臺灣東部海域)(臺灣東部海域)  2.0                 1         NaN   
4965   宜蘭縣政府東南方  30.3  公里 (位於宜蘭縣近海)(宜蘭縣近海)  3.0                 1         NaN   
4968    宜蘭縣政府東方  14.1  公里 (位於宜蘭縣近海)(宜蘭縣近海)  3.0                 1         NaN   
5013  花蓮縣政府東方  58.3  公里 (位於臺灣東部海域)(臺灣東部海域)  1.0                 1         NaN   
5025    花蓮縣政府東南方  4.0  公里 (位於花蓮縣近海)(花蓮縣近海)  4.0                 1         NaN   

      ...   Content Date Full IP   IP Longitude and Latitude Mobile Nation  \
4959  ...       NaN  NaN     NaN  NaN                    NaN    NaN    NaN   
4965  ...       NaN  NaN     NaN  NaN                    NaN    NaN    NaN   
4968  ...       NaN  NaN     NaN  NaN                    NaN    NaN    NaN   
5013  ...       NaN  NaN     NaN  NaN                    NaN    NaN    NaN   
5025  ...       NaN  NaN     NaN  NaN                    NaN    NaN    NaN   

     Push Title YouWin  
4959  NaN   NaN    NaN  
4965  NaN   NaN    NaN  
4968  NaN   NaN    NaN  
5013  NaN   NaN    NaN  
5025  NaN   NaN    NaN  

[5 rows x 23 columns]

### 進一步判斷一起地震其後有多少「地震文」

In [15]:
earthquake_index

[4959,
 4965,
 4968,
 5013,
 5025,
 5030,
 5046,
 5051,
 5065,
 5076,
 5186,
 5197,
 5199,
 5232,
 5248,
 5262,
 5266,
 5278,
 5305,
 5315,
 5319,
 5324,
 5361,
 5490,
 5508,
 5511,
 5538,
 5566,
 5605,
 5630,
 5644,
 5649,
 5655,
 5682,
 5704,
 5718,
 5731,
 5739,
 5758,
 5763,
 5769,
 5774,
 5792,
 5795,
 5802,
 5807,
 5809,
 5866,
 5874,
 5895,
 5901,
 5994,
 6035,
 6039,
 6064,
 6072,
 6078,
 6107,
 6119,
 6126,
 6176,
 6192,
 6232,
 6289,
 6291,
 6340,
 6367,
 6369,
 6400,
 6406,
 6464,
 6498,
 6519,
 6530,
 6563,
 6569,
 6619,
 6785,
 6797,
 6919,
 6927,
 6972,
 7018,
 7067,
 7095,
 7166,
 7174,
 7192,
 7239,
 7248,
 7311]

In [14]:
df_new = pd.DataFrame()
for i in earthquake_index:
    j = i #indicator for appending 地震文
    df_new = df_new.append(df.iloc[j], ignore_index=True)
    j+=1
    while df.iloc[j]["IsOfficialReport"]!=1: #地震報告後東西是貼文的話就要append進來
        df_new = df_new.append(df.iloc[j], ignore_index=True)
        j+=1 # keep appending 地震文 till 遇到offical report
len(df_new)

723

## 重拿一次氣象局地震報告的index

In [15]:
earthquake_index = []
for i in range(len(df_new)):
    if df_new.iloc[i]["IsOfficialReport"]==1:
        earthquake_index.append(i)
print(earthquake_index)

[0, 4, 7, 35, 44, 46, 52, 56, 64, 73, 75, 79, 81, 85, 91, 93, 96, 99, 104, 107, 110, 112, 143, 166, 180, 182, 184, 186, 191, 195, 199, 202, 204, 212, 234, 237, 240, 247, 250, 255, 260, 265, 268, 270, 276, 278, 280, 288, 293, 297, 299, 312, 318, 322, 324, 332, 338, 345, 357, 364, 397, 411, 440, 469, 471, 481, 488, 490, 493, 499, 505, 510, 515, 526, 530, 533, 540, 553, 556, 559, 561, 563, 577, 584, 593, 600, 604, 607, 654, 663, 712]


In [16]:
CenterLoc = []
for i in earthquake_index:
    j = i
    #if str(df_new.iloc[j]["位置"]) != "nan":
    CenterLoc.append(df_new.iloc[j]["位置"])
    j+=1
    try:
        while str(df_new.iloc[j]["位置"]) == "nan":
            CenterLoc.append(df_new.iloc[i]["位置"]) #如果是地震文，就把震央位置給填上 地震報告 的位置
            
            j+=1
    except:
        pass
len(CenterLoc)

723

In [17]:
# 同時處理震央及震度
CenterLoc = []
ShakingExtent = []
for i in earthquake_index:
    j = i
    CenterLoc.append(df_new.iloc[j]["位置"])
    ShakingExtent.append(df_new.iloc[i]["震度"])
    j+=1
    try:
        while str(df_new.iloc[j]["位置"]) == "nan":
            CenterLoc.append(df_new.iloc[i]["位置"]) #如果是地震文，就把震央位置給填上 地震報告 的位置
            ShakingExtent.append(df_new.iloc[i]["震度"])
            j+=1
    except:
        pass
len(CenterLoc)
len(ShakingExtent)

723

In [18]:
CenterLoc[0:10]

['宜蘭縣政府東方  74.2  公里 (位於臺灣東部海域)(臺灣東部海域)',
 '宜蘭縣政府東方  74.2  公里 (位於臺灣東部海域)(臺灣東部海域)',
 '宜蘭縣政府東方  74.2  公里 (位於臺灣東部海域)(臺灣東部海域)',
 '宜蘭縣政府東方  74.2  公里 (位於臺灣東部海域)(臺灣東部海域)',
 '宜蘭縣政府東南方  30.3  公里 (位於宜蘭縣近海)(宜蘭縣近海)',
 '宜蘭縣政府東南方  30.3  公里 (位於宜蘭縣近海)(宜蘭縣近海)',
 '宜蘭縣政府東南方  30.3  公里 (位於宜蘭縣近海)(宜蘭縣近海)',
 '宜蘭縣政府東方  14.1  公里 (位於宜蘭縣近海)(宜蘭縣近海)',
 '宜蘭縣政府東方  14.1  公里 (位於宜蘭縣近海)(宜蘭縣近海)',
 '宜蘭縣政府東方  14.1  公里 (位於宜蘭縣近海)(宜蘭縣近海)']

In [19]:
ShakingExtent[0:10]

[2.0, 2.0, 2.0, 2.0, 3.0, 3.0, 3.0, 3.0, 3.0, 3.0]

In [20]:
df_new["CenterLoc"] = CenterLoc
df_new["ShakingExtent"] = ShakingExtent

In [21]:
df_new.head()

Author      Board           City  \
0                  NaN        NaN            NaN   
1  DevNull (/dev/null)  Gossiping  Mountain View   
2    jackyuki1 (喵喵快來 )  Gossiping         Taipei   
3        Dihang (阿美阿美)  Gossiping         Taipei   
4                  NaN        NaN            NaN   

                                             Content  \
0                                                NaN   
1  發震時間：2019/05/08 23:48:47\n位置：北緯 24.81度，東經 122....   
2  為什麼每次地震完\n\n小弟想上ptt 都上不去\n\n大家手腳都好快....馬上就上ptt...   
3  我國東北的三寶是什麼？\n\n人參，貂皮，烏拉草\n\n還是\n\n人參，貂皮，鹿茸\n\n...   
4                                                NaN   

                       Date          Date_format           Full IP  \
0                       NaN  2019-05-08 23:48:47               NaN   
1  Wed May  8 23:49:28 2019  2019-05-08 23:49:28    35.236.191.107   
2  Wed May  8 23:56:55 2019  2019-05-08 23:56:55    112.104.89.169   
3  Thu May  9 00:43:35 2019  2019-05-09 00:43:35   114.136.232.233   
4                       NaN  2019-05-10 21:49:30               NaN   

                IP  IsOfficialReport  Latitude      ...        \
0              NaN               1.0    24.820      ...         
1   35.236.191.107               0.0    37.404      ...         
2   112.104.89.169               0.0    25.048      ...         
3  114.136.232.233               0.0    25.048      ...         
4              NaN               1.0    24.500      ...         

                  Title Unnamed: 0 YouWin  \
0                   NaN        NaN    NaN   
1      [爆卦] 東北外海地震 口牟~~     1983.0    1.0   
2  [問卦] 為什麼每次地震完ptt....     1982.0    0.0   
3  Re: [爆卦] 東北外海地震 口牟~~     1981.0    0.0   
4                   NaN        NaN    NaN   

                                     位置     深度       編號   規模   震度  \
0  宜蘭縣政府東方  74.2  公里 (位於臺灣東部海域)(臺灣東部海域)  122.7  小區域有感地震  5.1  2.0   
1                                   NaN    NaN      NaN  NaN  NaN   
2                                   NaN    NaN      NaN  NaN  NaN   
3                                   NaN    NaN      NaN  NaN  NaN   
4   宜蘭縣政府東南方  30.3  公里 (位於宜蘭縣近海)(宜蘭縣近海)   40.4  小區域有感地震  3.9  3.0   

                              CenterLoc  ShakingExtent  
0  宜蘭縣政府東方  74.2  公里 (位於臺灣東部海域)(臺灣東部海域)            2.0  
1  宜蘭縣政府東方  74.2  公里 (位於臺灣東部海域)(臺灣東部海域)            2.0  
2  宜蘭縣政府東方  74.2  公里 (位於臺灣東部海域)(臺灣東部海域)            2.0  
3  宜蘭縣政府東方  74.2  公里 (位於臺灣東部海域)(臺灣東部海域)            2.0  
4   宜蘭縣政府東南方  30.3  公里 (位於宜蘭縣近海)(宜蘭縣近海)            3.0  

[5 rows x 25 columns]

In [22]:
df_new = df_new[['Author', 'City', 'Content', 'Date_format', 'Full IP',
       'IP', 'IsOfficialReport', 'Latitude', 'Longitude',
        'Mobile', 'Nation', 'Push',
       'Title', 'YouWin', '位置', '深度', '編號', '規模', '震度', 'CenterLoc', 'ShakingExtent']]

In [23]:
df_new.head()

Author           City  \
0                  NaN            NaN   
1  DevNull (/dev/null)  Mountain View   
2    jackyuki1 (喵喵快來 )         Taipei   
3        Dihang (阿美阿美)         Taipei   
4                  NaN            NaN   

                                             Content          Date_format  \
0                                                NaN  2019-05-08 23:48:47   
1  發震時間：2019/05/08 23:48:47\n位置：北緯 24.81度，東經 122....  2019-05-08 23:49:28   
2  為什麼每次地震完\n\n小弟想上ptt 都上不去\n\n大家手腳都好快....馬上就上ptt...  2019-05-08 23:56:55   
3  我國東北的三寶是什麼？\n\n人參，貂皮，烏拉草\n\n還是\n\n人參，貂皮，鹿茸\n\n...  2019-05-09 00:43:35   
4                                                NaN  2019-05-10 21:49:30   

            Full IP               IP  IsOfficialReport  Latitude  Longitude  \
0               NaN              NaN               1.0    24.820    122.490   
1    35.236.191.107   35.236.191.107               0.0    37.404   -122.075   
2    112.104.89.169   112.104.89.169               0.0    25.048    121.532   
3   114.136.232.233  114.136.232.233               0.0    25.048    121.532   
4               NaN              NaN               1.0    24.500    121.920   

             Mobile      ...       \
0               NaN      ...        
1      Google Cloud      ...        
2    Digital United      ...        
3  Chunghwa Telecom      ...        
4               NaN      ...        

                                                Push                 Title  \
0                                                NaN                   NaN   
1  ['推 minirr: 幹 05/08 23:49', '推 cosmoswaves: 震 ...      [爆卦] 東北外海地震 口牟~~   
2  ['→ brianuser: 24小時掛著就不會上不去了 05/08 23:58', '',...  [問卦] 為什麼每次地震完ptt....   
3  ['→ gogoeggs: 肥宅 處男 小雞雞 05/09 00:44', '→ asked...  Re: [爆卦] 東北外海地震 口牟~~   
4                                                NaN                   NaN   

  YouWin                                    位置     深度       編號   規模   震度  \
0    NaN  宜蘭縣政府東方  74.2  公里 (位於臺灣東部海域)(臺灣東部海域)  122.7  小區域有感地震  5.1  2.0   
1    1.0                                   NaN    NaN      NaN  NaN  NaN   
2    0.0                                   NaN    NaN      NaN  NaN  NaN   
3    0.0                                   NaN    NaN      NaN  NaN  NaN   
4    NaN   宜蘭縣政府東南方  30.3  公里 (位於宜蘭縣近海)(宜蘭縣近海)   40.4  小區域有感地震  3.9  3.0   

                              CenterLoc ShakingExtent  
0  宜蘭縣政府東方  74.2  公里 (位於臺灣東部海域)(臺灣東部海域)           2.0  
1  宜蘭縣政府東方  74.2  公里 (位於臺灣東部海域)(臺灣東部海域)           2.0  
2  宜蘭縣政府東方  74.2  公里 (位於臺灣東部海域)(臺灣東部海域)           2.0  
3  宜蘭縣政府東方  74.2  公里 (位於臺灣東部海域)(臺灣東部海域)           2.0  
4   宜蘭縣政府東南方  30.3  公里 (位於宜蘭縣近海)(宜蘭縣近海)           3.0  

[5 rows x 21 columns]

### 計算時間差

In [26]:
import datetime
import time

In [24]:
earthquake_index = []
for i in range(len(df_new)):
    if df_new.iloc[i]["IsOfficialReport"]==1:
        earthquake_index.append(i)
len(earthquake_index)

91

In [27]:
# Set a list that contains 地震文與地震報告的時間差
TimeDifference = []
for j in range(len(df_new)):
    if df_new.iloc[j]["IsOfficialReport"] == 1: #如果是地震報告，則時間差為0
        dstr = df_new.iloc[j]["Date_format"] #找到時間標記
        d1 = datetime.datetime.strptime(dstr, "%Y-%m-%d %H:%M:%S") #Start Time 化為電腦的格式
        TimeDifference.append(0)
    else:
        dstr = df_new.iloc[j]["Date_format"] #找到時間標記
        d2 = datetime.datetime.strptime(dstr, "%Y-%m-%d %H:%M:%S") #地震文的發文時間化為電腦的格式
        d = d2-d1
        TimeDifference.append(d.total_seconds())

In [28]:
len(TimeDifference)

723

In [29]:
df_new["TimeDifference"] = TimeDifference

In [30]:
len(df_new[df_new["TimeDifference"]<600]) #先以十分鐘為基準：超過十分鐘才發文的都不算想搶地震文

462

### 其他地震失敗文與地震爆文的時間差

In [31]:
# Set a list that contains 地震失敗文與地震爆文的時間差
TimeDifference_Bao = []
for j in range(len(df_new)):
    if df_new.iloc[j]["YouWin"] == 1:
        dstr = df_new.iloc[j]["Date_format"] #找到時間標記
        d1 = datetime.datetime.strptime(dstr, "%Y-%m-%d %H:%M:%S") #Start Time 化為電腦的格式
        TimeDifference_Bao.append(0)
    else:
        dstr = df_new.iloc[j]["Date_format"] #找到時間標記
        d2 = datetime.datetime.strptime(dstr, "%Y-%m-%d %H:%M:%S") #地震文的發文時間化為電腦的格式
        d = d2-d1
        TimeDifference_Bao.append(d.total_seconds())

In [32]:
len(TimeDifference_Bao)

723

In [33]:
df_new["TimeDifference_Bao"] = TimeDifference_Bao

In [34]:
len(df_new[df_new["TimeDifference_Bao"]<600])

392

In [35]:
df_new[df_new["TimeDifference_Bao"]<600].head()

Author           City  \
0                  NaN            NaN   
1  DevNull (/dev/null)  Mountain View   
2    jackyuki1 (喵喵快來 )         Taipei   
5  tony77998 (偷尼柒柒玖玖捌)         Taipei   
8             ros (螺絲)         Taipei   

                                             Content          Date_format  \
0                                                NaN  2019-05-08 23:48:47   
1  發震時間：2019/05/08 23:48:47\n位置：北緯 24.81度，東經 122....  2019-05-08 23:49:28   
2  為什麼每次地震完\n\n小弟想上ptt 都上不去\n\n大家手腳都好快....馬上就上ptt...  2019-05-08 23:56:55   
5  幹  板橋有吧\n\n7777777777777777   快去開門阿  塊陶阿   補個字...  2019-05-10 21:50:07   
8  幹你娘我在6樓快暈死了\n\n最近沒有下雨怎麼會地震\n\n辦公室好多人手機發出警報\n\n...  2019-05-13 16:55:49   

           Full IP              IP  IsOfficialReport  Latitude  Longitude  \
0              NaN             NaN               1.0    24.820    122.490   
1   35.236.191.107  35.236.191.107               0.0    37.404   -122.075   
2   112.104.89.169  112.104.89.169               0.0    25.048    121.532   
5    58.115.130.21   58.115.130.21               0.0    25.048    121.532   
8    36.231.45.115   36.231.45.115               0.0    25.048    121.532   

                     Mobile        ...         YouWin  \
0                       NaN        ...            NaN   
1              Google Cloud        ...            1.0   
2            Digital United        ...            0.0   
5  Hoshin Multimedia Center        ...            1.0   
8                     HiNet        ...            1.0   

                                     位置     深度       編號   規模   震度  \
0  宜蘭縣政府東方  74.2  公里 (位於臺灣東部海域)(臺灣東部海域)  122.7  小區域有感地震  5.1  2.0   
1                                   NaN    NaN      NaN  NaN  NaN   
2                                   NaN    NaN      NaN  NaN  NaN   
5                                   NaN    NaN      NaN  NaN  NaN   
8                                   NaN    NaN      NaN  NaN  NaN   

                              CenterLoc  ShakingExtent  TimeDifference  \
0  宜蘭縣政府東方  74.2  公里 (位於臺灣東部海域)(臺灣東部海域)            2.0             0.0   
1  宜蘭縣政府東方  74.2  公里 (位於臺灣東部海域)(臺灣東部海域)            2.0            41.0   
2  宜蘭縣政府東方  74.2  公里 (位於臺灣東部海域)(臺灣東部海域)            2.0           488.0   
5   宜蘭縣政府東南方  30.3  公里 (位於宜蘭縣近海)(宜蘭縣近海)            3.0            37.0   
8    宜蘭縣政府東方  14.1  公里 (位於宜蘭縣近海)(宜蘭縣近海)            3.0            28.0   

  TimeDifference_Bao  
0        -35018893.0  
1                0.0  
2              447.0  
5                0.0  
8                0.0  

[5 rows x 23 columns]

# 計算距離

台灣地區若以 *北緯24度 東經121度* 即南投埔里附近為原點

則：

* 往東往西差一經度約差101公里

* 往南往北差一經度約差110公里

In [36]:
# Set a list that contains 地震文與地震報告的距離差
Distance = []
for j in range(len(df_new)):
    if df_new.iloc[j]["IsOfficialReport"] == 1: #如果是地震報告，則距離差為0
        X0 = df_new.iloc[j]["Longitude"] #找到經緯度
        Y0 = df_new.iloc[j]["Latitude"]  #當作原點
        Distance.append(0)
    else:
        X = df_new.iloc[j]["Longitude"] #找到經緯度
        Y = df_new.iloc[j]["Latitude"]
        d = np.sqrt( (101*(X-X0))**2 + (110*(Y-Y0))**2 )
        Distance.append(d)

In [37]:
df_new["Distance"] = Distance

In [39]:
df_new[df_new["TimeDifference"]<600].head()

Author           City  \
0                  NaN            NaN   
1  DevNull (/dev/null)  Mountain View   
2    jackyuki1 (喵喵快來 )         Taipei   
4                  NaN            NaN   
5  tony77998 (偷尼柒柒玖玖捌)         Taipei   

                                             Content          Date_format  \
0                                                NaN  2019-05-08 23:48:47   
1  發震時間：2019/05/08 23:48:47\n位置：北緯 24.81度，東經 122....  2019-05-08 23:49:28   
2  為什麼每次地震完\n\n小弟想上ptt 都上不去\n\n大家手腳都好快....馬上就上ptt...  2019-05-08 23:56:55   
4                                                NaN  2019-05-10 21:49:30   
5  幹  板橋有吧\n\n7777777777777777   快去開門阿  塊陶阿   補個字...  2019-05-10 21:50:07   

           Full IP              IP  IsOfficialReport  Latitude  Longitude  \
0              NaN             NaN               1.0    24.820    122.490   
1   35.236.191.107  35.236.191.107               0.0    37.404   -122.075   
2   112.104.89.169  112.104.89.169               0.0    25.048    121.532   
4              NaN             NaN               1.0    24.500    121.920   
5    58.115.130.21   58.115.130.21               0.0    25.048    121.532   

                     Mobile      ...       \
0                       NaN      ...        
1              Google Cloud      ...        
2            Digital United      ...        
4                       NaN      ...        
5  Hoshin Multimedia Center      ...        

                                     位置     深度       編號   規模   震度  \
0  宜蘭縣政府東方  74.2  公里 (位於臺灣東部海域)(臺灣東部海域)  122.7  小區域有感地震  5.1  2.0   
1                                   NaN    NaN      NaN  NaN  NaN   
2                                   NaN    NaN      NaN  NaN  NaN   
4   宜蘭縣政府東南方  30.3  公里 (位於宜蘭縣近海)(宜蘭縣近海)   40.4  小區域有感地震  3.9  3.0   
5                                   NaN    NaN      NaN  NaN  NaN   

                              CenterLoc ShakingExtent  TimeDifference  \
0  宜蘭縣政府東方  74.2  公里 (位於臺灣東部海域)(臺灣東部海域)           2.0             0.0   
1  宜蘭縣政府東方  74.2  公里 (位於臺灣東部海域)(臺灣東部海域)           2.0            41.0   
2  宜蘭縣政府東方  74.2  公里 (位於臺灣東部海域)(臺灣東部海域)           2.0           488.0   
4   宜蘭縣政府東南方  30.3  公里 (位於宜蘭縣近海)(宜蘭縣近海)           3.0             0.0   
5   宜蘭縣政府東南方  30.3  公里 (位於宜蘭縣近海)(宜蘭縣近海)           3.0            37.0   

   TimeDifference_Bao      Distance  
0         -35018893.0      0.000000  
1                 0.0  24739.820786  
2               447.0     99.955575  
4            165602.0      0.000000  
5                 0.0     71.898385  

[5 rows x 24 columns]

# 簡化

In [40]:
# 簡化 只看那些時間差在十分鐘內的
df = df_new[df_new["TimeDifference"]<600]

# 新增三個columns

1. 發文的縣市是不是震央所在地

2. 發文的時間是不是深夜或凌晨

3. 所位在的地方的人口密度

In [41]:
df.columns

Index(['Author', 'City', 'Content', 'Date_format', 'Full IP', 'IP',
       'IsOfficialReport', 'Latitude', 'Longitude', 'Mobile', 'Nation', 'Push',
       'Title', 'YouWin', '位置', '深度', '編號', '規模', '震度', 'CenterLoc',
       'ShakingExtent', 'TimeDifference', 'TimeDifference_Bao', 'Distance'],
      dtype='object')

In [42]:
set(df_new["位置"].dropna())

{'南投縣政府南偏東方  29.0  公里 (位於南投縣竹山鎮)(南投縣竹山鎮)',
 '南投縣政府南偏東方  29.2  公里 (位於南投縣竹山鎮)(南投縣竹山鎮)',
 '南投縣政府東偏北方  33.2  公里 (位於南投縣埔里鎮)(南投縣埔里鎮)',
 '嘉義縣政府南偏西方  9.7  公里 (位於嘉義縣鹿草鄉)(嘉義縣鹿草鄉)',
 '嘉義縣政府東偏北方  44.9  公里 (位於嘉義縣梅山鄉)(嘉義縣梅山鄉)',
 '嘉義縣政府東偏南方  22.7  公里 (位於嘉義縣中埔鄉)(嘉義縣中埔鄉)',
 '嘉義縣政府東偏南方  8.9  公里 (位於嘉義縣水上鄉)(嘉義縣水上鄉)',
 '基隆市政府東方  44.7  公里 (位於臺灣東北部海域)(臺灣東北部海域)',
 '宜蘭縣政府南偏東方  31.1  公里 (位於宜蘭縣近海)(宜蘭縣近海)',
 '宜蘭縣政府南偏東方  35.9  公里 (位於臺灣東部海域)(臺灣東部海域)',
 '宜蘭縣政府南偏東方  38.1  公里 (位於臺灣東部海域)(臺灣東部海域)',
 '宜蘭縣政府南偏東方  43.0  公里 (位於臺灣東部海域)(臺灣東部海域)',
 '宜蘭縣政府南偏東方  46.3  公里 (位於臺灣東部海域)(臺灣東部海域)',
 '宜蘭縣政府南偏西方  23.4  公里 (位於宜蘭縣南澳鄉)(宜蘭縣南澳鄉)',
 '宜蘭縣政府南偏西方  28.7  公里 (位於宜蘭縣南澳鄉)(宜蘭縣南澳鄉)',
 '宜蘭縣政府南方  44.6  公里 (位於宜蘭縣南澳鄉)(宜蘭縣南澳鄉)',
 '宜蘭縣政府東偏北方  14.1  公里 (位於宜蘭縣近海)(宜蘭縣近海)',
 '宜蘭縣政府東偏北方  14.5  公里 (位於宜蘭縣近海)(宜蘭縣近海)',
 '宜蘭縣政府東偏北方  30.3  公里 (位於臺灣東部海域)(臺灣東部海域)',
 '宜蘭縣政府東偏北方  32.3  公里 (位於臺灣東部海域)(臺灣東部海域)',
 '宜蘭縣政府東偏北方  8.4  公里 (位於宜蘭縣近海)(宜蘭縣近海)',
 '宜蘭縣政府東偏南方  24.9  公里 (位於臺灣東部海域)(臺灣東部海域)',
 '宜蘭縣政府東偏南方  44.4  公里 (位於臺灣東部海域)(臺灣東部海域)',
 '宜蘭縣政府東偏南方  69.6  公里 (位

In [43]:
set(df["City"].dropna())

{'Ashburn',
 'Chang-hua',
 'Chiayi City',
 'Douliu',
 'Hsinchu',
 'Hualien City',
 'Kaohsiung City',
 'Magong',
 'Miaoli',
 'Mountain View',
 'Nantou City',
 'Parsippany',
 'Petaling Jaya',
 'Pingtung City',
 'Taichung',
 'Tainan City',
 'Taipei',
 'Tampa',
 'Taoyuan City',
 'Taoyuan District',
 'Yilan',
 'Zhongli District',
 'Zhubei',
 '經度：121.000\u3000緯度：23.500'}

In [44]:
# 建字典
District_dict = {
 'Ashburn'       : "Ashburn, Virginia, U.S.A.",
 'Chang-hua'     : "彰化",
 'Chiayi City'   : "嘉義",
 'Douliu'        : "斗六",
 'Hsinchu'       : "新竹",
 'Hualien City'  : "花蓮",
 'Kaohsiung City': "高雄",
 'Magong'        : "馬公",
 'Miaoli'        : "苗栗",
 'Nantou City'   : "南投",
 'Parsippany'    : "Parsippany-Troy Hills, New Jersey, U.S.A.",
 'Petaling Jaya' : "Petaling Jaya, Petaling District, Selangor, Malaysia", 
 'Pingtung City' : "屏東",
 'Taichung'      : "台中",
 'Tainan City'   : "台南",
 'Taipei'        : "臺北",
 'Tampa'         : "Tampa, Florida, U.S.A",
 'Taoyuan City'  : "桃園",
 'Taoyuan District' : "桃園",
 'Yilan'         : "宜蘭",
 'Zhongli District' :"中壢",
 'Zhubei'        : "竹北"
}

In [45]:
City_Translated = []
for district in df["City"]:
    if district in District_dict:
        City_Translated.append(District_dict[district])
    else:
        City_Translated.append("NaN")

In [46]:
df["City_Translated"] = City_Translated

/Applications/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [49]:
InCenter = []
for i in range(len(df)):
    if str(df.iloc[i]["City_Translated"]) in str(df.iloc[i]["CenterLoc"]):
        InCenter.append(1)
    else:
        InCenter.append(0)
print(InCenter)
print(len(InCenter))

[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 

In [50]:
df["InCenter"] = InCenter

/Applications/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [51]:
df[df["InCenter"]==1].head()

Author            City  \
58        HAHAcomet (值得信任的彗星小天使)  Kaohsiung City   
76           cake51640 (uuuu4ni)     Chiayi City   
111        online135 (98分美元宇宙星塵)    Hualien City   
113      openbestbook (吃蘿蔔不吐蘿蔔皮)          Taipei   
114  imjungyi923 (邊緣BL寫手 小旋風伊咪達)          Taipei   

                                               Content          Date_format  \
58   \n\n\n\n剛剛地震發生前 八卦人數大概是14XXX\n\n地震後的現在 馬上增加到19...  2019-05-31 23:59:26   
76                                             幹地震\n\n  2019-06-26 05:09:08   
111  地震!!!\n\n\n※註:有電視或媒體有報導者，請勿使用爆卦!\n無重大八卦請勿使用此分類...  2019-08-06 09:20:36   
113  \n哞哞哞哞哞哞\n\n小精靈縮4台北市附近發生有感地震  哞~~\n五樓快承認翻身跌下床\...  2019-08-07 01:01:01   
114                   新北中和有感\n\n雖然只小晃一下\n\n還以為是我頭暈\n\n  2019-08-07 01:01:11   

                   Full IP               IP  IsOfficialReport  Latitude  \
58           114.39.50.137    114.39.50.137               0.0    22.615   
76     111.242.114.80 (臺灣)   111.242.114.80               0.0    23.463   
111    114.45.170.130 (臺灣)   114.45.170.130               0.0    23.981   
113   101.137.196.181 (臺灣)  101.137.196.181               0.0    25.048   
114      49.217.192.7 (臺灣)     49.217.192.7               0.0    25.048   

     Longitude                Mobile   ...      編號  規模  震度  \
58     120.314                 HiNet   ...     NaN NaN NaN   
76     120.582                 HiNet   ...     NaN NaN NaN   
111    121.612                 HiNet   ...     NaN NaN NaN   
113    121.532  Asia Pacific Telecom   ...     NaN NaN NaN   
114    121.532         TWN Broadband   ...     NaN NaN NaN   

                                  CenterLoc ShakingExtent  TimeDifference  \
58    高雄市政府東北方  43.1  公里 (位於高雄市美濃區)(高雄市美濃區)           4.0           368.0   
76   嘉義縣政府東偏南方  22.7  公里 (位於嘉義縣中埔鄉)(嘉義縣中埔鄉)           3.0            34.0   
111  花蓮縣政府南偏西方  36.4  公里 (位於花蓮縣豐濱鄉)(花蓮縣豐濱鄉)           5.0            48.0   
113    臺北市政府北方  13.0  公里 (位於臺北市士林區)(臺北市士林區)           3.0            20.0   
114    臺北市政府北方  13.0  公里 (位於臺北市士林區)(臺北市士林區)           3.0            30.0   

    TimeDifference_Bao   Distance  City_Translated InCenter  
58               349.0  42.981210               高雄        1  
76                 0.0  12.326736               嘉義        1  
111                0.0  35.198233               花蓮        1  
113                0.0  11.366333               臺北        1  
114               10.0  11.366333               臺北        1  

[5 rows x 26 columns]

### 尖峰與離峰時段：另外兩個dummies

尖峰時段就設定在peak的前後兩個小時 

* 22:00~2:00

* 13:00~17:00

離峰時間設定在

* 4:00~8:00

* 18:00~20:00

In [52]:
df.iloc[1]["Date_format"]

'2019-05-08 23:49:28'

In [55]:
dstr = df.iloc[31]["Date_format"]
d1 = datetime.datetime.strptime(dstr, "%Y-%m-%d %H:%M:%S")
d1.timetuple()
d1.timetuple()[3] #表示 "時"

17

In [56]:
# Set a list that contains dummies for peak & offpeak
IsPeak = []
OffPeak = []
for j in range(len(df)):
    dstr = df.iloc[j]["Date_format"] #找到時間標記
    d1 = datetime.datetime.strptime(dstr, "%Y-%m-%d %H:%M:%S") #Start Time 化為電腦的格式
    hour = d1.timetuple()[3] #拿到使用者是在幾點發文的
    
    if hour in [22, 23, 0, 1, 13, 14, 15, 16]:
        IsPeak.append(1)
    else:
        IsPeak.append(0)
    
    if hour in [4, 5, 6, 7, 18, 19]:
        OffPeak.append(1)
    else:
        OffPeak.append(0)
print(len(IsPeak), len(OffPeak))

462 462


In [57]:
df["IsPeak"] = IsPeak
df["OffPeak"] = OffPeak

/Applications/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.
/Applications/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [58]:
df.head()

Author           City  \
0                  NaN            NaN   
1  DevNull (/dev/null)  Mountain View   
2    jackyuki1 (喵喵快來 )         Taipei   
4                  NaN            NaN   
5  tony77998 (偷尼柒柒玖玖捌)         Taipei   

                                             Content          Date_format  \
0                                                NaN  2019-05-08 23:48:47   
1  發震時間：2019/05/08 23:48:47\n位置：北緯 24.81度，東經 122....  2019-05-08 23:49:28   
2  為什麼每次地震完\n\n小弟想上ptt 都上不去\n\n大家手腳都好快....馬上就上ptt...  2019-05-08 23:56:55   
4                                                NaN  2019-05-10 21:49:30   
5  幹  板橋有吧\n\n7777777777777777   快去開門阿  塊陶阿   補個字...  2019-05-10 21:50:07   

           Full IP              IP  IsOfficialReport  Latitude  Longitude  \
0              NaN             NaN               1.0    24.820    122.490   
1   35.236.191.107  35.236.191.107               0.0    37.404   -122.075   
2   112.104.89.169  112.104.89.169               0.0    25.048    121.532   
4              NaN             NaN               1.0    24.500    121.920   
5    58.115.130.21   58.115.130.21               0.0    25.048    121.532   

                     Mobile   ...     震度  \
0                       NaN   ...    2.0   
1              Google Cloud   ...    NaN   
2            Digital United   ...    NaN   
4                       NaN   ...    3.0   
5  Hoshin Multimedia Center   ...    NaN   

                              CenterLoc ShakingExtent  TimeDifference  \
0  宜蘭縣政府東方  74.2  公里 (位於臺灣東部海域)(臺灣東部海域)           2.0             0.0   
1  宜蘭縣政府東方  74.2  公里 (位於臺灣東部海域)(臺灣東部海域)           2.0            41.0   
2  宜蘭縣政府東方  74.2  公里 (位於臺灣東部海域)(臺灣東部海域)           2.0           488.0   
4   宜蘭縣政府東南方  30.3  公里 (位於宜蘭縣近海)(宜蘭縣近海)           3.0             0.0   
5   宜蘭縣政府東南方  30.3  公里 (位於宜蘭縣近海)(宜蘭縣近海)           3.0            37.0   

  TimeDifference_Bao      Distance City_Translated  InCenter  IsPeak OffPeak  
0        -35018893.0      0.000000             NaN         0       1       0  
1                0.0  24739.820786             NaN         0       1       0  
2              447.0     99.955575              臺北         0       1       0  
4           165602.0      0.000000             NaN         0       0       0  
5                0.0     71.898385              臺北         0       0       0  

[5 rows x 28 columns]

## 人口密度

利用臺灣行政區面積表及臺灣行政區人口列表的資料計算

https://zh.wikipedia.org/wiki/臺灣行政區人口密度表

In [59]:
# 建字典 人口密度 單位：人/每平方公里
popDensity_dict = {
    "彰化" : 1182.02,
    "嘉義" : 263.38,
    "斗六" : 526.38, #使用縣市層級，i.e.雲林縣的人口密度
    "新竹" : 4318.72,
    "花蓮" : 70.32,
    "高雄" : 939.35,
    "馬公" : 827.83, #使用澎湖縣
    "苗栗" : 299.04,
    "南投" : 120.03,
    "屏東" : 294.05,
    "台中" : 1271.61,
    "台南" : 857,
    "臺北" : 9662.94,
    "桃園" : 1848.28,
    "宜蘭" : 211.78,
    "中壢" : 1848.28,#使用桃園
    "竹北" : 4318.72 #使用新竹
}

In [60]:
PopDensity = []
for city in df["City_Translated"]:
    if city in popDensity_dict.keys():
        PopDensity.append(popDensity_dict[city])
    else:
        PopDensity.append(float("NaN"))
len(PopDensity)

462

In [61]:
df["PopDensity"] = PopDensity

/Applications/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [62]:
df.head()

Author           City  \
0                  NaN            NaN   
1  DevNull (/dev/null)  Mountain View   
2    jackyuki1 (喵喵快來 )         Taipei   
4                  NaN            NaN   
5  tony77998 (偷尼柒柒玖玖捌)         Taipei   

                                             Content          Date_format  \
0                                                NaN  2019-05-08 23:48:47   
1  發震時間：2019/05/08 23:48:47\n位置：北緯 24.81度，東經 122....  2019-05-08 23:49:28   
2  為什麼每次地震完\n\n小弟想上ptt 都上不去\n\n大家手腳都好快....馬上就上ptt...  2019-05-08 23:56:55   
4                                                NaN  2019-05-10 21:49:30   
5  幹  板橋有吧\n\n7777777777777777   快去開門阿  塊陶阿   補個字...  2019-05-10 21:50:07   

           Full IP              IP  IsOfficialReport  Latitude  Longitude  \
0              NaN             NaN               1.0    24.820    122.490   
1   35.236.191.107  35.236.191.107               0.0    37.404   -122.075   
2   112.104.89.169  112.104.89.169               0.0    25.048    121.532   
4              NaN             NaN               1.0    24.500    121.920   
5    58.115.130.21   58.115.130.21               0.0    25.048    121.532   

                     Mobile    ...                                 CenterLoc  \
0                       NaN    ...      宜蘭縣政府東方  74.2  公里 (位於臺灣東部海域)(臺灣東部海域)   
1              Google Cloud    ...      宜蘭縣政府東方  74.2  公里 (位於臺灣東部海域)(臺灣東部海域)   
2            Digital United    ...      宜蘭縣政府東方  74.2  公里 (位於臺灣東部海域)(臺灣東部海域)   
4                       NaN    ...       宜蘭縣政府東南方  30.3  公里 (位於宜蘭縣近海)(宜蘭縣近海)   
5  Hoshin Multimedia Center    ...       宜蘭縣政府東南方  30.3  公里 (位於宜蘭縣近海)(宜蘭縣近海)   

  ShakingExtent TimeDifference  TimeDifference_Bao      Distance  \
0           2.0            0.0         -35018893.0      0.000000   
1           2.0           41.0                 0.0  24739.820786   
2           2.0          488.0               447.0     99.955575   
4           3.0            0.0            165602.0      0.000000   
5           3.0           37.0                 0.0     71.898385   

   City_Translated InCenter  IsPeak  OffPeak PopDensity  
0              NaN        0       1        0        NaN  
1              NaN        0       1        0        NaN  
2               臺北        0       1        0    9662.94  
4              NaN        0       0        0        NaN  
5               臺北        0       0        0    9662.94  

[5 rows x 29 columns]

# 拿掉official report，因為它們只是參照點，並非觀察值

In [63]:
df_new = df

In [64]:
# 拿掉official report，因為它們只是參照點，並非觀察值
df = df[df["IsOfficialReport"]==0]

# 看一些簡單的統計量

In [65]:
np.mean(df["TimeDifference"]) #平均發文間隔

127.38814016172506

In [66]:
np.mean(df[df["YouWin"]==1]["TimeDifference"]) #地震爆文平均發文間隔

46.55140186915888

In [67]:
#df["YouWin"]
np.mean(df["YouWin"]) #搶到爆文的比率

0.2884097035040431

In [68]:
np.mean(df["Distance"]) #平均距離

451.6240886074431

In [69]:
np.mean(df[df["YouWin"]==1]["Distance"]) #那些搶到爆文的人平均而言離震央的距離

842.461655696156

In [70]:
# 位處震央的人的比例
np.mean(df["InCenter"])

0.12398921832884097

In [71]:
np.mean(df["IsPeak"]) #有36%的人在尖峰時段上ptt

0.3692722371967655

In [72]:
np.mean(df["OffPeak"]) #有36%的人在尖峰時段上ptt

0.3261455525606469

In [73]:
np.mean(df[df["IsPeak"]==1]["YouWin"]) #在尖峰時段搶到爆文的比例

0.3284671532846715

In [74]:
np.mean(df[df["OffPeak"]==1]["YouWin"]) #在離峰時段搶到爆文的比例

0.1652892561983471

In [75]:
np.mean(df[df["YouWin"]==1]["PopDensity"]) #那些搶到爆文的人所在的城市平均而言的人口密度

4616.575212765959

# 拿掉不要的columns以到R跑迴歸

In [76]:
df.columns

Index(['Author', 'City', 'Content', 'Date_format', 'Full IP', 'IP',
       'IsOfficialReport', 'Latitude', 'Longitude', 'Mobile', 'Nation', 'Push',
       'Title', 'YouWin', '位置', '深度', '編號', '規模', '震度', 'CenterLoc',
       'ShakingExtent', 'TimeDifference', 'TimeDifference_Bao', 'Distance',
       'City_Translated', 'InCenter', 'IsPeak', 'OffPeak', 'PopDensity'],
      dtype='object')

### 變數說明

'Date_format' : 時間

'YouWin' : 是否是地震爆文

'TimeDifference': 發文時間與中央氣象局的地震報告標注的地震發生時間的差

'Distance': IP位址反查得到的經緯度與地震震央經緯度的距離

'ShakingExtent':震度

'InCenter':是否為在震央所在縣市

'City_Translated':發文者所在位置

'CenterLoc':震央

'IsPeak' : 是否在尖峰時段發文

'OffPeak': 是否在離峰時段發文

'PopDensity' : 發文者所在的縣市的人口密度



'Author':作者ID

'Content':發文內容

'Mobile':使用的電信業者

'Push':推文內容

'Title':發文標題

In [77]:
df_R = df[['Date_format','YouWin','TimeDifference', 'TimeDifference_Bao', 'Distance', 'ShakingExtent',
           'InCenter', 'City_Translated', 'CenterLoc', 'IsPeak', 'OffPeak', 'PopDensity',
           'Title', 'Author', 'Content', 'Mobile', 'Push']]

# Save file

In [94]:
df_R.to_csv("combined_forR.csv", encoding="utf-8")

# 畫地圖的資料

In [78]:
df_new.columns

Index(['Author', 'City', 'Content', 'Date_format', 'Full IP', 'IP',
       'IsOfficialReport', 'Latitude', 'Longitude', 'Mobile', 'Nation', 'Push',
       'Title', 'YouWin', '位置', '深度', '編號', '規模', '震度', 'CenterLoc',
       'ShakingExtent', 'TimeDifference', 'TimeDifference_Bao', 'Distance',
       'City_Translated', 'InCenter', 'IsPeak', 'OffPeak', 'PopDensity'],
      dtype='object')

In [79]:
df_graph = df_new[["Date_format", "IsOfficialReport", "Latitude", "Longitude", "震度", "YouWin"]]

In [80]:
df_graph.head()

Date_format  IsOfficialReport  Latitude  Longitude   震度  YouWin
0  2019-05-08 23:48:47               1.0    24.820    122.490  2.0     NaN
1  2019-05-08 23:49:28               0.0    37.404   -122.075  NaN     1.0
2  2019-05-08 23:56:55               0.0    25.048    121.532  NaN     0.0
4  2019-05-10 21:49:30               1.0    24.500    121.920  3.0     NaN
5  2019-05-10 21:50:07               0.0    25.048    121.532  NaN     1.0

In [98]:
df_graph.to_csv("Data_graphing.csv", encoding="utf-8")